File: sdks/kmm/shared/src/commonMain/kotlin/com/truesight/sdk/TrueSight.kt
package com.truesight.sdk

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.datetime.Clock
import kotlinx.datetime.TimeZone
import kotlinx.datetime.toLocalDateTime
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonPrimitive

object TrueSight {

    private const val TAG = "TrueSight"
    internal const val SDK_VERSION = "1.0.0"

    private var initialized = false
    private lateinit var config: Config
    private lateinit var eventQueue: EventQueue
    private lateinit var networkClient: NetworkClient
    private lateinit var flushScheduler: FlushScheduler
    private lateinit var anonymousIdManager: AnonymousIdManager
    private lateinit var deviceContextCollector: DeviceContextCollector
    private lateinit var logger: Logger

    private var userId: String? = null
    private var anonymousId: String = ""
    private var mobileNumber: String? = null
    private var email: String? = null
    private var traits: Map<String, Any?> = emptyMap()

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
    private val mutex = Mutex()
    private var flushing = false

    fun init(config: Config) {
        init(
            config = config,
            eventQueue = EventQueue(),
            networkClient = NetworkClient(),
            flushScheduler = FlushScheduler(),
            anonymousIdManager = AnonymousIdManager(),
            deviceContextCollector = DeviceContextCollector(),
            logger = Logger()
        )
    }

    internal fun init(
        config: Config,
        eventQueue: EventQueue,
        networkClient: NetworkClient,
        flushScheduler: FlushScheduler,
        anonymousIdManager: AnonymousIdManager,
        deviceContextCollector: DeviceContextCollector,
        logger: Logger
    ) {
        if (initialized) {
            logger.warn(TAG, "TrueSight SDK already initialized")
            return
        }

        this.config = config
        this.eventQueue = eventQueue
        this.networkClient = networkClient
        this.flushScheduler = flushScheduler
        this.anonymousIdManager = anonymousIdManager
        this.deviceContextCollector = deviceContextCollector
        this.logger = logger

        anonymousId = anonymousIdManager.getOrCreate()

        flushScheduler.start(config.flushInterval) {
            performFlush()
        }

        initialized = true

        if (config.debug) {
            logger.debug(TAG, "TrueSight SDK initialized with endpoint: ${config.endpoint}")
            logger.debug(TAG, "Anonymous ID: $anonymousId")
        }
    }

    fun track(eventName: String, properties: Map<String, Any?> = emptyMap()) {
        ensureInitialized()
        val event = buildEvent(
            eventName = eventName,
            eventType = EventType.TRACK,
            properties = properties
        )
        enqueueEvent(event)
    }

    fun identify(userId: String, traits: Map<String, Any?> = emptyMap()) {
        ensureInitialized()
        this.userId = userId
        this.traits = traits

        // Auto-promote mobile_number and email from traits
        traits["mobile_number"]?.let { value ->
            if (value is String) {
                val cleaned = value.replace(Regex("[^0-9]"), "")
                if (cleaned.length == 10) {
                    mobileNumber = cleaned
                }
            }
        }
        traits["email"]?.let { value ->
            if (value is String && value.contains("@")) {
                email = value
            }
        }

        val event = buildEvent(
            eventName = "identify",
            eventType = EventType.IDENTIFY,
            properties = traits
        )
        enqueueEvent(event)

        if (config.debug) {
            logger.debug(TAG, "User identified: $userId")
        }
    }

    fun screen(screenName: String, properties: Map<String, Any?> = emptyMap()) {
        ensureInitialized()
        val mergedProperties = properties.toMutableMap().apply {
            put("screen_name", screenName)
        }
        val event = buildEvent(
            eventName = screenName,
            eventType = EventType.SCREEN,
            properties = mergedProperties
        )
        enqueueEvent(event)
    }

    fun flush() {
        ensureInitialized()
        scope.launch {
            performFlush()
        }
    }

    fun reset() {
        ensureInitialized()
        userId = null
        mobileNumber = null
        email = null
        traits = emptyMap()
        anonymousId = anonymousIdManager.reset()
        scope.launch {
            eventQueue.clear()
        }
        if (config.debug) {
            logger.debug(TAG, "SDK state reset. New anonymous ID: $anonymousId")
        }
    }

    fun setMobileNumber(number: String) {
        ensureInitialized()
        val cleaned = number.replace(Regex("[^0-9]"), "")
        if (cleaned.length != 10) {
            logger.warn(TAG, "Invalid mobile number: must be exactly 10 digits, got ${cleaned.length}")
            return
        }
        mobileNumber = cleaned
        if (config.debug) {
            logger.debug(TAG, "Mobile number set: ${cleaned.take(3)}***${cleaned.takeLast(2)}")
        }
    }

    fun setEmail(email: String) {
        ensureInitialized()
        if (!email.contains("@") || !email.contains(".")) {
            logger.warn(TAG, "Invalid email format: $email")
            return
        }
        this.email = email
        if (config.debug) {
            logger.debug(TAG, "Email set: $email")
        }
    }

    // -- Internal helpers --

    internal fun isInitialized(): Boolean = initialized

    internal fun getCurrentUserId(): String? = userId

    internal fun getAnonymousId(): String = anonymousId

    internal fun getMobileNumber(): String? = mobileNumber

    internal fun getEmail(): String? = email

    private fun ensureInitialized() {
        check(initialized) { "TrueSight SDK not initialized. Call TrueSight.init(config) first." }
    }

    private fun buildEvent(
        eventName: String,
        eventType: EventType,
        properties: Map<String, Any?>
    ): EventModel {
        val now = Clock.System.now()
        val timestamp = now.toLocalDateTime(TimeZone.UTC).toString() + "Z"

        return EventModel(
            eventId = generateUUID(),
            eventName = eventName,
            eventType = eventType,
            userId = userId,
            anonymousId = anonymousId,
            mobileNumber = mobileNumber,
            email = email,
            clientTimestamp = timestamp,
            properties = convertProperties(properties),
            context = deviceContextCollector.collect()
        )
    }

    private fun convertProperties(properties: Map<String, Any?>): Map<String, JsonElement> {
        return properties.mapValues { (_, value) ->
            when (value) {
                null -> JsonNull
                is String -> JsonPrimitive(value)
                is Number -> JsonPrimitive(value)
                is Boolean -> JsonPrimitive(value)
                else -> JsonPrimitive(value.toString())
            }
        }
    }

    private fun enqueueEvent(event: EventModel) {
        scope.launch {
            try {
                val serialized = Serializer.serializeEvent(event)
                if (serialized.length > config.maxEventSize) {
                    logger.warn(TAG, "Event exceeds max size (${serialized.length} > ${config.maxEventSize}), dropping")
                    return@launch
                }

                val currentSize = eventQueue.size()
                if (currentSize >= config.maxQueueSize) {
                    logger.warn(TAG, "Event queue full ($currentSize events), dropping oldest")
                    val oldest = eventQueue.dequeue(1)
                    if (oldest.isNotEmpty()) {
                        eventQueue.remove(oldest.map { it.eventId })
                    }
                }

                eventQueue.enqueue(event)

                if (config.debug) {
                    logger.debug(TAG, "Event enqueued: ${event.eventName} (${event.eventType})")
                }

                // Auto-flush if batch size reached
                if (eventQueue.size() >= config.maxBatchSize) {
                    performFlush()
                }
            } catch (e: Exception) {
                logger.error(TAG, "Failed to enqueue event: ${event.eventName}", e)
            }
        }
    }

    private suspend fun performFlush() {
        mutex.withLock {
            if (flushing) return
            flushing = true
        }

        try {
            while (true) {
                val events = eventQueue.dequeue(config.maxBatchSize)
                if (events.isEmpty()) break

                val now = Clock.System.now()
                val sentAt = now.toLocalDateTime(TimeZone.UTC).toString() + "Z"
                val payload = BatchPayload(batch = events, sentAt = sentAt)

                when (val result = networkClient.sendBatch(config.endpoint, config.apiKey, payload)) {
                    is NetworkResult.Success -> {
                        eventQueue.remove(events.map { it.eventId })
                        if (config.debug) {
                            logger.debug(TAG, "Flush successful: ${result.accepted} events accepted")
                        }
                    }
                    is NetworkResult.ClientError -> {
                        logger.error(TAG, "Client error ${result.code}: ${result.message}")
                        // Drop events on client error (4xx) - they won't succeed on retry
                        eventQueue.remove(events.map { it.eventId })
                        break
                    }
                    is NetworkResult.ServerError -> {
                        logger.warn(TAG, "Server error ${result.code}: ${result.message}, will retry")
                        // Keep events in queue for retry
                        break
                    }
                    is NetworkResult.NetworkFailure -> {
                        logger.warn(TAG, "Network failure, will retry: ${result.cause.message}")
                        // Keep events in queue for retry
                        break
                    }
                }
            }
        } catch (e: Exception) {
            logger.error(TAG, "Flush failed unexpectedly", e)
        } finally {
            mutex.withLock {
                flushing = false
            }
        }
    }

    /**
     * Resets internal state for testing purposes.
     */
    internal fun tearDown() {
        if (initialized) {
            flushScheduler.stop()
        }
        initialized = false
        userId = null
        mobileNumber = null
        email = null
        traits = emptyMap()
        anonymousId = ""
        flushing = false
    }
}

/**
 * Platform-specific UUID generation.
 */
expect fun generateUUID(): String

File: sdks/web/src/truesight.ts
import type { Config } from './config.js';
import { buildConfig } from './config.js';
import type { TrueSightEvent, EventType } from './event-model.js';
import { EventQueue } from './event-queue.js';
import { FlushScheduler } from './flush-scheduler.js';
import { getDeviceContext, clearDeviceContextCache } from './device-context.js';
import {
  getOrCreateAnonymousId,
  resetAnonymousId,
} from './anonymous-id.js';
import { readUserId, writeUserId, clearUserId } from './user-id.js';
import { clearKeyCache } from './encryption.js';
import { logger } from './logger.js';
import { SessionManager } from './session-manager.js';
import { AutoTrackManager } from './auto/auto-track-manager.js';

export class TrueSightSDK {
  private initialized = false;
  private config!: Config;
  private queue!: EventQueue;
  private scheduler!: FlushScheduler;
  private sessionManager!: SessionManager;
  private autoTrackManager: AutoTrackManager | null = null;

  private userId: string | null = null;
  private anonymousId: string = '';
  private mobileNumber: string | null = null;
  private email: string | null = null;
  private traits: Record<string, unknown> = {};

  async init(initConfig: {
    apiKey: string;
    endpoint: string;
    options?: Partial<Config>;
  }): Promise<void> {
    if (this.initialized) {
      logger.warn('TrueSight already initialized');
      return;
    }

    if (!initConfig.apiKey) {
      throw new Error('apiKey is required');
    }
    if (!initConfig.endpoint) {
      throw new Error('endpoint is required');
    }

    this.config = buildConfig(initConfig);
    logger.setDebug(this.config.debug);

    // Initialize anonymous ID
    this.anonymousId = await getOrCreateAnonymousId(this.config.apiKey);

    // Restore persisted user ID
    this.userId = await readUserId(this.config.apiKey);

    // Open IndexedDB queue
    this.queue = new EventQueue(this.config.apiKey, this.config.maxQueueHard);
    await this.queue.open();

    // Start flush scheduler
    this.scheduler = new FlushScheduler(this.config, this.queue);
    this.scheduler.start();

    // Initialize session manager
    this.sessionManager = new SessionManager(
      this.config.sessionTimeout,
      (eventName, properties) => {
        void this.enqueueEvent('track', eventName, properties).catch((error) => {
          logger.warn(`Failed to enqueue session event "${eventName}": ${String(error)}`);
        });
      }
    );

    this.initialized = true;
    logger.info('TrueSight SDK initialized');

    // Initialize auto-tracking (must be after initialized=true since trackers call track())
    this.autoTrackManager = new AutoTrackManager(
      this.config.autoTrack,
      (eventName, properties) => {
        void this.track(eventName, properties).catch((error) => {
          logger.warn(`Failed to auto-track event "${eventName}": ${String(error)}`);
        });
      },
      (screenName, properties) => {
        void this.screen(screenName, properties).catch((error) => {
          logger.warn(`Failed to auto-track screen "${screenName}": ${String(error)}`);
        });
      }
    );
    this.autoTrackManager.start();
  }

  async track(
    eventName: string,
    properties: Record<string, unknown> = {}
  ): Promise<void> {
    this.ensureInitialized();

    if (!eventName) {
      logger.warn('track() called with empty event name');
      return;
    }

    await this.enqueueEvent('track', eventName, properties);
  }

  async identify(
    userId: string,
    traits: Record<string, unknown> = {}
  ): Promise<void> {
    this.ensureInitialized();

    if (!userId) {
      logger.warn('identify() called with empty userId');
      return;
    }

    this.userId = userId;
    this.traits = { ...this.traits, ...traits };

    // Persist user ID for cross-reload continuity
    await writeUserId(userId, this.config.apiKey);

    // Auto-promote mobile_number and email from traits
    if (
      typeof traits.mobile_number === 'string' &&
      traits.mobile_number
    ) {
      this.mobileNumber = traits.mobile_number;
    }
    if (typeof traits.email === 'string' && traits.email) {
      this.email = traits.email;
    }

    // Backfill queued events that have no user_id
    await this.queue.backfillUserId(userId);

    await this.enqueueEvent('identify', 'identify', traits);
  }

  async screen(
    screenName: string,
    properties: Record<string, unknown> = {}
  ): Promise<void> {
    this.ensureInitialized();

    if (!screenName) {
      logger.warn('screen() called with empty screen name');
      return;
    }

    await this.enqueueEvent('screen', '$screen', {
      screen_name: screenName,
      ...properties,
    });
  }

  async flush(): Promise<void> {
    this.ensureInitialized();
    await this.scheduler.flush();
  }

  async trackExperiment(
    flagName: string,
    variant: string,
    properties: Record<string, unknown> = {}
  ): Promise<void> {
    await this.track('$experiment_exposure', {
      flag_name: flagName,
      variant,
      ...properties,
    });
  }

  async reset(): Promise<void> {
    this.ensureInitialized();

    this.userId = null;
    this.mobileNumber = null;
    this.email = null;
    this.traits = {};

    // Clear persisted user ID
    clearUserId();

    // Reset session
    if (this.sessionManager) {
      this.sessionManager.reset();
    }

    // Generate new anonymous ID
    this.anonymousId = await resetAnonymousId(this.config.apiKey);

    logger.debug('SDK state reset');
  }

  setMobileNumber(number: string): void {
    this.ensureInitialized();

    // Validate 10 digits
    const digits = number.replace(/\D/g, '');
    if (digits.length !== 10) {
      logger.warn(
        `Invalid mobile number: expected 10 digits, got ${digits.length}`
      );
      return;
    }

    this.mobileNumber = digits;
    logger.debug('Mobile number set');
  }

  setEmail(email: string): void {
    this.ensureInitialized();
    this.email = email;
    logger.debug('Email set');
  }

  /** Check if SDK is initialized (useful for testing). */
  isInitialized(): boolean {
    return this.initialized;
  }

  /** Get current user ID (useful for testing). */
  getUserId(): string | null {
    return this.userId;
  }

  /** Get current anonymous ID (useful for testing). */
  getAnonymousId(): string {
    return this.anonymousId;
  }

  /** Get current mobile number (useful for testing). */
  getMobileNumber(): string | null {
    return this.mobileNumber;
  }

  /** Get current email (useful for testing). */
  getEmail(): string | null {
    return this.email;
  }

  /** Destroy SDK instance (stop scheduler, close DB). */
  destroy(): void {
    if (this.autoTrackManager) {
      this.autoTrackManager.stop();
      this.autoTrackManager = null;
    }
    if (this.sessionManager) {
      this.sessionManager.destroy();
    }
    if (this.scheduler) {
      this.scheduler.stop();
    }
    if (this.queue) {
      this.queue.close();
    }
    this.initialized = false;
    this.userId = null;
    this.anonymousId = '';
    this.mobileNumber = null;
    this.email = null;
    this.traits = {};
    clearKeyCache();
    clearDeviceContextCache();
    logger.debug('SDK destroyed');
  }

  private ensureInitialized(): void {
    if (!this.initialized) {
      throw new Error(
        'TrueSight SDK not initialized. Call init() first.'
      );
    }
  }

  private async enqueueEvent(
    eventType: EventType,
    eventName: string,
    properties: Record<string, unknown>
  ): Promise<void> {
    const context = await getDeviceContext();

    // Touch session activity + get session ID
    const sessionId = this.sessionManager
      ? (this.sessionManager.touchActivity(), this.sessionManager.getSessionId())
      : null;

    const event: TrueSightEvent = {
      event_id: crypto.randomUUID(),
      event_name: eventName,
      event_type: eventType,
      user_id: this.userId,
      anonymous_id: this.anonymousId,
      mobile_number: this.mobileNumber,
      email: this.email,
      session_id: sessionId,
      client_timestamp: new Date().toISOString(),
      properties,
      context,
    };

    // Check event size
    const serialized = JSON.stringify(event);
    if (serialized.length > this.config.maxEventSize) {
      logger.warn(
        `Event exceeds maxEventSize (${serialized.length} > ${this.config.maxEventSize}), dropping`
      );
      return;
    }

    // Check queue soft limit
    const queueSize = await this.queue.size();
    if (queueSize >= this.config.maxQueueSize) {
      logger.warn(
        `Queue at soft limit (${queueSize}/${this.config.maxQueueSize}), triggering flush`
      );
      this.scheduler.scheduleFlush();
    }

    await this.queue.enqueue(event);
    logger.debug(`Event queued: ${eventType} - ${eventName}`);
  }
}
